### ISSUE TITLE
1 - Introduction: 1.3 Kinds of Machine Learning Problems

### ISSUE THREAD

**Issue by JeremyLoh** [JeremyLoh/d2l#4]
https://d2l.ai/chapter_introduction/index.html#kinds-of-machine-learning-problems

# 1.3.1 Supervised Learning

Supervised learning describes tasks where we are given a dataset containing both features and labels, and we are tasked to produce a model to predict the labels given input features
Each feature - label pair is called an example
The supervision comes into play because for choosing the parameters, we (the supervisors) provide the model with a dataset consisting of labelled examples

In probabilistic terms, we are typically interested in estimating the conditional probability of a label given input features

Many important tasks can be described as estimating the probability of something unknown given a particular set of available data
- Predict cancer vs non-cancer, given computer tomography image
- Predict the correct translation of language in French, given sentence in English
- Predict stock price next month based on this month financial reporting data

Supervised learning problems can be described as "predicting the labels given input features". But they can have tons of modelling decisions, depending on type, size and quantity of the inputs and outputs. e.g. we use different models to process sequences of arbitrary lengths and for processing fixed-length vector representations

Informally, the supervised learning process looks like
1. Grab a big collection of examples where the features are known and select from them a random subset. Acquirer ground truth labels for each. Sometimes, these data are already collected or we need human annotators to label the data (e.g. assign images to categories). The inputs and corresponding labels are the **training set**
2. Feed the training dataset into supervised learning algorithm. It takes in dataset as input and as output, returns the learned model
3. Feed previously unseen inputs to the learned model, using its outputs as predictions of the corresponding label

![Image](https://user-images.githubusercontent.com/38811217/258624103-e03ae86c-3235-4e16-8a5f-c577f2a6753f.png)

# 1.3.1.1 Regression
Simplest supervised learning task to wrap your head around
e.g. a set of data obtained about home sales. Each row is a different house and each column represent some relevant attribute (e.g. square footage of house, number of bathrooms, walking time to center of town)
Each example would be a specific house, and corresponding feature vector would be one row in the table
Fixed length feature vectors like this are essential for most classic machine learning algorithms

What makes a problem a regression is the form of the target
e.g. You want to estimate the fair market value of a house, given some features like above
Data might consist of historical home listings and the labels might be the observed sales prices
**When labels take on arbitrary numerical values (even within some interval), we call this a regression problem**
The goal is to produce a model whose predictions closely approximate the actual label values

Lots of practical problems can be described as regression problems
- Predicting the rating a user will assign a movie
- Predict the length of stay of patients in the hospital

Good rule of thumb is any **how much? or how many** problems suggest regression
- how many hours will this surgery take?
- how much rainfall will this town have in next 6 hours? 

Sometimes we can't exactly predict the observed values. There is some distance between our predictions and the observed values
**In most of our chapters, we will focus on minimizing the squared error loss function. This loss corresponds to the assumption that our data were corrupted by Gaussian noise**

# 1.3.1.2 Classification
e.g. taking photo of a cheque and an app that will scan letters. It will need to determine each handwritten character and which label it belongs to, among a list of known character set.

In classification, we want our model to look at features E.g. pixel values in image, and then predict which category (sometimes called a class) among some discrete set of options, an example belongs
e.g. For handwritten digits, we have ten classes, from digit 0 to 9.
**The simplest form of classification is where there is only two classes, binary classification**
- e.g. images of animals and we have labels { cat, dog }

In regression, we sought a regressor to output a numerical value
In classification, we seek a classifier, whose output is the predicted class assignment

It can be hard to optimize a model that can only output a hard categorical assignment (e.g. either "cat" or "dog")
In these cases, it is easier to express our model in the language of probabilities
Given features of an example, our model assign a probability to each possible class. e.g. probability output that image is a cat is 0.9 (90%)

**Multiclass classification is when we have more than two possible classes for our problem**
- e.g. hand written character recognition { 0, 1, 2, ... 9, a, b, c, ... }

The common loss function for regression problem is to minimize the squared error loss function
The common loss function for classification problems is called **cross-entropy**

Classification can get more complicated than binary or multiclass classification
e.g. there are some variants of classification addressing hierarchically structured classes. In this case, not all errors are equal. e.g. we might prefer to misclassify to a related class rather than a distant class. This is referred to as **hierarchical classification**.
- e.g. the model would pay a huge penalty if it confused a poodle for a dinosaur

#1.3.1.3 Tagging
e.g. We might run in trouble for classification problems if there are multiple animals present in the image
Might want to give the model the option of saying that the image depicts a cat, a dog, a donkey and a rooster

The problem of learning to predict classes that are not mutually exclusive (2 or more events that cannot happen simultaneously) is called **multi-label classification**

Auto-tagging problems are best described as multi-label classification problems. e.g. tags that people might give to posts on technical blog, "machine learning", "technology", "linux"
**Typically, tags will exhibit some correlation structure.** 
- Posts about cloud computing are likely to mention "AWS" and posts about "machine learning" are likely to mention "GPUs"

Sometimes, tagging problems draw on enormous label sets

# 1.3.1.4 Search
In information retrieval, we often impose rankings over sets of items
e.g. web search, where the goal is **not to** determine whether a particular page is relevant for a query, but which, among a set of relevant results should be shown more prominently to a particular user

# 1.3.1.5 Recommender Systems
Related to search and ranking
Goal is to display a set of relevant items to the user. The main difference is the personalization to specific users in the context of recommender systems

In some cases, customers provide explicit feedback, communicating how much they like a product. (e.g. rating and review on Amazon, Goodreads).
In other cases, they will provide **implicit **feedback**, e.g skipping titles on playlist

In simplest formulations, these systems will be trained to estimate some score, e.g. some star rating or probability that a given user will purchase a particular item

Recommendation system built on top of predictive models suffer some serious conceptual flaws
- We only observe censored feedback. Users preferentially rate movies that they feel strongly about. There are lesser 3 star ratings compared to 1 star or 5 star ratings
- Current purchase habits are often a result of recommendation algorithm already in place, but learning algorithms do not take into account this detail. There can be feedback loops where a recommender system preferentially pushes an item that is then taken to be better (due to more purchases), and it is then recommended more frequently.

# 1.3.1.6 Sequence Learning
So far, we have looked at problems where we have some fixed number of inputs and produce a fixed number of outputs
e.g. predict housing price given a fixed set of features (e.g. square footage of house, number of bedrooms)
In these cases, **once our model is trained, after each test example is fed into our model, it is immediately forgotten**. We assumed that successive observations **were independent and thus there was no need to hold on to this context**

But how do we deal with video snippets?? Each snippet might contain a different number of frames. Our guess on what is happening for each frame might be stronger if we take into account the previous or succeeding frames. Same for language

e.g. Machine translation: task of ingesting sentences in some source language and predicting their translations in another language

e.g. medicine, where we want to predict and monitor patients in intensive care unit, to fire alerts whenever their risk of dying in the next 24 hours exceed some threshold. Here, we won't throw away everything that we know about the patient history every hour. We will make predictions based on only the most recent measurements

These are examples of sequence learning.
Require a model to either ingest sequences of inputs or to emit sequences of output (or both)

**Sequence-to-sequence learning** considers problems where inputs and outputs both consist of variable-length sequences
- e.g. machine translation, and speech to text transcription

 # 1.3.2 Unsupervised and Self-Supervised Learning
The previous examples focused on supervised learning, where we feed the model a giant dataset containing both features and corresponding label values

Unsupervised learning is where we get handed a large dataset and are asked to answer questions. Can be any question, limited to our creativity
- e.g. Clustering - can we find a small number of prototypes that accurately summarize the data? e.g. given a set of photos, can we group them into landscape photos, pictures of dogs, etc.

# 1.3.3 Interacting with an Environment
We have not discussed where data actually comes from, or what happens when a machine learning model generates an output

This is because supervised learning and unsupervised learning do not address these issues in a very sophisticated way
- In either case, we grab a big pile of data upfront, then set our pattern recognition machines in motion without ever interacting with the environment again
- **Offline learning** - where all the learning takes place after the algorithm is disconnected from the environment

Benefit of offline learning
- Can worry about pattern recognition in isolation, no need to worry about interactions with a dynamic environment

![Image](https://user-images.githubusercontent.com/38811217/258628314-9f2fcc08-1cc3-4773-a3fd-8e5cba664292.png)

# 1.3.4 Reinforcement Learning
A framework for posing learning problems in which an agent interacts with an environment

For developing an agent that interacts with an environment and takes actions
e.g. applications such as robotics, dialogue systems, AI for video games

Deep reinforcement learning, which applies deep learning to reinforcement learning problems

Reinforcement learning gives a very general statement of a problem, in which an agent interacts with an environment over a series of time steps
At each time step, the agent receives some observation from the environment and must choose an action that is subsequently transmitted back to the environment via some mechanism (sometimes called an actuator)
Finally, the agent receives a reward from the environment

The behavior of a reinforcement learning agent is governed by a policy
A policy is a function that maps from observations of the environment to actions.
The goal of reinforcement learning is to produce good policies

![Image](https://user-images.githubusercontent.com/38811217/258631702-a98489f6-77ee-4e45-8c5d-4ff7fa3d5abf.png)

In reinforcement learning, we do not assume that for each observation, the environment tells us the optimal action
In general, we just get some reward
The environment may not even tell us which actions led to the reward
- e.g. chess, the only real reward is at the end of the game. when we either win (reward of 1) or lose (-1)

Reinforcement learners also have to deal with the potential problem of partial observability. The current observation might not tell you everything about your current state

At any given point, reinforcement learners might know of one good policy, but there might be many other better policies that the agent has never tried. The reinforcement learner must constantly choose whether to exploit the best (currently) known strategy as a policy, or to explore the space of strategies, potentially giving up some short-run reward in exchange for knowledge.

The general reinforcement learning problem is a very general setting. Actions affect subsequent observations. Rewards are only observed corresponding to the chosen actions. The environment may be either fully or partially observed. Accounting for all this complexity at once may ask too much of researchers. Moreover, not every practical problem exhibits all this complexity. As a result, researchers have studied a number of special cases of reinforcement learning problems.

**When the environment is fully observed, we call the reinforcement learning problem a Markov Decision Process**

**When the state does not depend on the previous actions, we call the problem a contextual bandit problem**

**When there is no state, just a set of available actions with initially unknown rewards, the problem is the classic multi-armed bandit problem**

The alternation of linear and nonlinear processing units, often referred to as layers.

The use of the chain rule (also known as backpropagation) for adjusting parameters in the entire network at once.

# 1.5 The Road to Deep Learning
Much of this changed with the availability of large amounts of data, due to the World Wide Web, the advent of companies serving hundreds of millions of users online, a dissemination of cheap, high-quality sensors, cheap data storage (Kryder’s law), and cheap computation (Moore’s law). In particular, the landscape of computation in deep learning was revolutionized by advances in GPUs, which were originally engineered for computer gaming. Suddenly algorithms and models that seemed computationally infeasible became relevant (and vice versa)

# 1.7 The Essence of Deep Learning
Deep learning is the subset of machine learning concerned with models based on many layered neural networks
It is deep in the sense that its models learn many layers of transformations

Arguably, all machine learning has many layers of computation, the first consisting of feature processing steps. What differentiates deep learning is that the operations learned at each of the many layers of representations are learned jointly from data.

Arguably the most significant commonality in deep learning methods is end-to-end training. That is, rather than assembling a system based on components that are individually tuned, one builds the system and then tunes their performance jointly
For instance, in computer vision scientists used to separate the process of feature engineering from the process of building machine learning models
